### PR TITLE
Refactor src/uartdev for async and code size

### PR DESCRIPTION
**Pull request by laurensvalk**
The current iodev and uartdev code will only ever be used for LEGO uart devices, so we can reduce the level of abstraction and complexity considerably. More generic devices would be handled by a port interface like proposed in https://github.com/pybricks/pybricks-micropython/pull/159.

The main reason for doing is to prepare for handling sensors in an async way compatible with #166. Now, we can make a single synchronous call to set the mode/data (and raise a relevant error as needed), and await data to be ready in the same way.

This also reduces the code size by about 570 bytes despite adding new functionality like buffered lpf2 writes, which should further reduce the (otherwise needed) size to implement async for sensors.

The final commit with buffered data writes is perhaps the more controversial one, as it gets a bit complicated. We could consider doing without that one and check if it's needed when we get to the async part.

### CODE REVIEW

**Review by dlech — COMMENTED** [review]
Looks good.

It would be nice if we could also extend the uartdev tests to get more coverage of the new code too.

**Comment by laurensvalk** [comment]
Todos:

-------

> It seems like we should be checking for this in the contiki process instead of here, so that we don't have to keep calling `pbio_iodev_is_ready()` to keep things moving along.

_Originally posted by @dlech in https://github.com/pybricks/pybricks-micropython/pull/169#discussion_r1199176769_
            
-----

File: pybricks/pupdevices/pb_type_pupdevices_colordistancesensor.c [hunk 1]
```diff
@@ -14,64 +14,62 @@
...
+        return PBIO_IODEV_MODE_PUP_COLOR_DISTANCE_SENSOR__AMBI; // blue
     }
+    return PBIO_IODEV_MODE_PUP_COLOR_DISTANCE_SENSOR__REFLT; // red
+}
 
-    pb_device_set_values(pbdev, PBIO_IODEV_MODE_PUP_COLOR_DISTANCE_SENSOR__COL_O, &color, 1);
+STATIC void pupdevices_ColorDistanceSensor_light_on(void *context, const pbio_color_hsv_t *hsv) {
+    pbio_iodev_t *iodev = context;
+    uint8_t *data;
+    pup_device_get_data(iodev, get_mode_for_color(hsv), &data);
```
[the remark this answers is not in the document]
Comment (laurensvalk) [reply pybricks/pupdevices/pb_type_pupdevices_colordistancesensor.c:72]: That's included:

```c
void pb_pup_device_get_data(pbio_iodev_t *iodev, uint8_t mode, uint8_t **data) {
    pb_assert(pbio_iodev_set_mode(iodev, mode));
    pb_pup_device_wait_ready(iodev);
    pb_assert(pbio_iodev_get_data(iodev, mode, data));
}
```

**Comment by laurensvalk** [comment]
There was a note on naming in one of the review comments.

Since `uartdev` `iodev` and `ioport` in their current forms are only for LEGO devices, maybe we could consider:

- `ludev`: LEGO UART Device (now `uartdev`)
- `lpdev`: LEGO Passive Device (now `ioport_lpf2`).

We could drop the `iodev` abstraction that tries to combine the two. This could perhaps save some more space since it drops mode info of passive devices that is mostly unused.

Names like `iodev` and `ioport` may be more suited for generic ports interfaces like in #159.

We could defer this until we actually get to #159 though.

**Comment by laurensvalk** [comment]
- [ ] Assert ID in get/set calls.

One limitation of the previous version of the code and carried over now, is that we only assert the device ID when instantiating the user device class.

In theory you could instantiate the class, unplug the sensor, plug in something else, and then do a read/write.

Now that everything is consolidated to just a few generic get/set calls, it may be a bit cheaper than before to assert not only the mode but also the ID.

**Comment by laurensvalk** [comment]
To make it safer, instead of 
```c
pbio_error_t pbio_iodev_get_data(pbio_iodev_t *iodev, uint8_t mode, void **data)
```

We could do:

```c
pbio_error_t pbio_iodev_get_data(pbio_iodev_t *iodev, uint8_t mode, void *data)
```

And let it explicitly unpack the data based on its type in the mode info.

**Comment by laurensvalk** [comment]
It is currently probably in a shape where I can start completing the (separate) async PR.

Would you be interested in taking this pull request further in the mean time?

**Comment by dlech** [comment]
> * `lpdev`: LEGO Passive Device (now `ioport_lpf2`).

`ioport_lpf2` is mainly the device connection manager for the I/O ports, so should not be renamed to lpdev. But we should be able to drop the iodev layer and just say "if detected device id on the ioport is ... then poke ioport directly else raise ENODEV" for passive devices kind of like we already do for motors.

**Comment by laurensvalk** [comment]
> > * `lpdev`: LEGO Passive Device (now `ioport_lpf2`).
> 
> `ioport_lpf2` is mainly the device connection manager for the I/O ports, so should not be renamed to lpdev. But we should be able to drop the iodev layer and just say "if detected device id on the ioport is ... then poke ioport directly else raise ENODEV" for passive devices kind of like we already do for motors.

Sounds good. I've been working on exactly this, let's see what we can save here...

**Comment by laurensvalk** [comment]
Thanks for the review and additional improvements!

Since several other PRs hinge on this, I've gone ahead and merged it. We can continue to polish in a follow up PR as needed.

**Pull request merged**